Reworked the whole code for being compatible with latest PyMISP

obert01 opened the pull request:
Reworked the whole code for being compatible with latest PyMISP >= 2.4.113.

It has not been easily possible to keep compatibility with older versions.

But the code has a few more improvements:
- Only fetch the list of available MISP tags once
- Fixed the check for existing MISP tag: a MISP tag is only added if it is not already present in the MISP event. 
- The same has been done at the attribute level to prevent PyMISP from generating errors. This part can even be optimized by doing all the stuff in memory. For the moment, a request is made to MISP to search for an existing similar attribute.

obert01 committed:
Reworked the whole code for being compatible with latest PyMISP >= 2.4.113.

It has not been easily possible to keep compatibility with older versions.

Quyrean commented:
Hi,

Thanks, I am having the same problem.  I tried using this code but get the following error.  I installed the latest pymisp,  any ideas?

```
git fetch origin pull/35/head:pr35
git checkout pr35
sudo pip install pymisp
sudo pip install .
otx-misp --otx ***** --server "*****" --misp ***** -vvvvvvv --dedup-titles
Connection to MISP instance: *****
Error: Cannot connect to MISP instance, unknown exception: 'module' object has no attribute 'ExpandedPyMISP'

```

obert01 commented:
Hello,

Are you sure you are using python 3, version 3.6 at least?

Maybe use pip3 instead of pip for installing. 

What is your pymisp version? (cf. pip3 list) 

Hope that helps

Quyrean commented:
@obert01 Thank you.  I updated to python 3 and that did the trick.  otx-misp ran for quite a while after that and entered many events.  Then it ended with 

  Error: 'NoneType' object has no attribute 'read'

have you seen this error before?
thanks!

obert01 commented:
I have seen this error for the first time today in my logs. But it has appeared only once, while I am pulling AlienVault OTX every 4 hours. I don't think it comes from the part of the code I have rewritten, as I haven't any call to a "read" method. 

If I understand well, this comes from the OTX connection part. This probably happens when the AlienVault OTX server does not answer quicly enough. 

Don't hesitate to run otx-misp with the --stop-on-error=no option. You may also wrap your call to otx-misp in a shell script, which may rerun otx-misp while there are errors (while the return value is not 0).

Quyrean commented:
@obert01 That works.  Thank you.  I hope this code can be merged in soon.

TheDr1ver commented on src/otx_misp/__init__.py:
```diff
@@ -255,36 +253,43 @@ def get_tag_name(complete):
-                if result['message'] == "No matches.":
-                    event = misp.new_event(distribution, threat_level, analysis, event_name, date=event_date,
-                                           published=publish)
+            result = misp.search(eventinfo=event_name, metadata=True)
```
~~See #30 PR - you'll want to search with pipes escaped due to the way the MISP backend handles them oddly.~~

~~...Strike that, this appears to work fine without escaping the pipe, so I guess MISP fixed something on the backend. However I'm still getting duplicates if I select `author=True` and `dedup_titles=True` so I'm wondering if it has something to do with non-deduped events being created as Python objects whereas the result of this search comes back as a JSON blob unless `pythonify=True` is set.~~

...nevermind - it was a problem with my test instance. Everything seems to be working as it should.

milannshrestha commented:
@obert01 hey man, im having following error:
```
/otx_misp/src/otx_misp/otx/otx_misp# otx-misp -o xxxxxxxxxxxxxxxxxxxxxxxxxx -s https://misp.xxxxx.xxxx-m xxxxxxxxxxxxxxxxxxxxxxxx-a -u --discover-tags --to-ids --distribution connected --author-tag --bulk-tag OTX --dedup-titles
Error: 'TypeError' object has no attribute 'message'```
```
Could you help me with this, running py3.

obert01 commented:
Hello @milannshrestha 

Just importing OTX event in a fresh MISP instance, but unfortunately, I can't reproduce your error. 

Could you please try executing the program again with verbose logging, e.g. adding the -vvvv option in your command line?

Thanks

spasaintk commented:
Quick reply for anyone else landing on this pull request. The TypeError is indeed fixed by applying this patch. If you still run into the error, you might not have downloaded and installed otx-misp with this fix correctly.

chrisinmtown commented:
This PR has been open about 14 months.  Has @gcrahay abandoned this project?

chrisinmtown commented on src/otx_misp/__init__.py:
```diff
@@ -60,32 +68,17 @@ def tag_event(misp, event, tag):
             tag_id = exist_tag['id']
-            if 'EventTag' in event:
-                for evt_tag in event['EventTag']:
+            if 'Tag' in event:
```
EventTag is certainly wrong, but I believe you must first extract the event dict using key 'Event', and *then* look for the tags dict using key 'Tag'. Here's the first part of a pretty-print of the event object in this function courtesy of the python debugger: 

```
(Pdb) pp(event)
{'Event': {'Attribute': [{'Galaxy': [],
                          'ShadowAttribute': [],
                          'category': 'External analysis',
                          'comment': '',
                          'deleted': False,
                          'disable_correlation': False,
    ...deleted...
           'ShadowAttribute': [],
           'Tag': [{'colour': '#ffffff',
                    'exportable': True,
                    'hide_tag': False,
                    'id': '3',
                    'local': 0,
                    'name': 'foo:white',
   ...deleted...
```

TheDr1ver committed:
fix api path

I know this repo is long-dead and the owner hasn't maintained it in years, but I'm still using it. A couple weeks ago AlienVault changed their API slightly and @shreyaganvir kindly mentioned the fix in issue #43 of the original repo.

obert01 committed:
Merge pull request #1 from TheDr1ver/patch-4

fix api path
